src/cdk/datepicker/datepicker-input.ts

```diff
@@ -0,0 +1,361 @@
+
+  /** Emits when the disabled state has changed. */
+  _disabledChange = new EventEmitter<boolean>();
+
+  /** Implemented as part of ControlValueAccessor. */
+  _onTouched = () => {};
+
+  /** Implemented as part of ControlValueAccessor. */
+  protected _controlValAccOnChange: (value: any) => void = () => {};
+
+  /** Implemented as part of ControlValueAccessor. */
+  protected _validatorOnChange = () => {};
+
+  /** Implemented for datepicker CDK subscription. */
+  private _cdkDatepickerSubscription = Subscription.EMPTY;
+
+  /** Implemented for datepicker locale subscription. */
+  private _localeSubscription = Subscription.EMPTY;
+
+  /** The native input element to which this directive is attached. */
+  protected _inputElement: HTMLInputElement;
+
+  /** The form control validator for whether the input parses. */
+  private _parseCdkValidator: ValidatorFn = (): ValidationErrors | null => {
+    return this._lastValueValid ?
+        null : {'cdkDatepickerParse': {'text': this._elementRef.nativeElement.value}};
+  }
+
+  /** The form control validator for the min date. */
+  private _minCdkValidator: ValidatorFn = (control: AbstractControl): ValidationErrors | null => {
+    const controlValue = this._getValidDateOrNull(this._dateAdapter.deserialize(control.value));
+    return (!this.min || !controlValue ||
+        this._dateAdapter.compareDate(this.min, controlValue) <= 0) ?
+        null : {'cdkDatepickerMin': {'min': this.min, 'actual': controlValue}};
+  }
+
+  /** The form control validator for the max date. */
+  private _maxCdkValidator: ValidatorFn = (control: AbstractControl): ValidationErrors | null => {
+    const controlValue = this._getValidDateOrNull(this._dateAdapter.deserialize(control.value));
+    return (!this.max || !controlValue ||
+        this._dateAdapter.compareDate(this.max, controlValue) >= 0) ?
+        null : {'cdkDatepickerMax': {'max': this.max, 'actual': controlValue}};
+  }
+
+  /** The form control validator for the date filter. */
+  private _filterCdkValidator: ValidatorFn = (control: AbstractControl):
+      ValidationErrors | null => {
+    const controlValue = this._getValidDateOrNull(this._dateAdapter.deserialize(control.value));
+    return !this._dateFilter || !controlValue || this._dateFilter(controlValue) ?
+        null : {'cdkDatepickerFilter': true};
+  }
+
+  /** The combined form control validator for this input. */
+  private _cdkValidator: ValidatorFn | null =
+      Validators.compose(
+          [this._parseCdkValidator, this._minCdkValidator, this._maxCdkValidator,
+              this._filterCdkValidator]);
+
+  /** Whether the last value set on the input was valid. */
+  protected _lastValueValid = false;
+
+  /** Element reference associated with this input. */
+  protected _elementRef: ElementRef;
+
+  /** DateAdapter associated with this input. */
+  _dateAdapter: DateAdapter<D>;
+
+  /** Constructor for the datepicker input component. */
+  constructor() {
```

Comment:
Don't you need to inject things here? `DateAdapter`, `ElementRef`?

Reply:
The DateAdapter is used in the `src/lib/datepicker/datepicker-input.ts` file as well. I created a protected variable in the CDK instead of injecting so that when the material datepicker-input is extended from the CDK datepicker-input, the dateAdapter would be used by both. I will inject `ElementRef`.

Reply:
You can do it in the constructor:

```ts
constructor(protected dateAdapter: DateAdapter<D>, ...) {}
```

then in the subclass you just pass it to the super constructor:

```ts
constructor(dateAdapter: DateAdapter<D>, ...) {
  super(dateAdapter, ...);
}
```

Reply:
I was going to approach it this way as well, but I didn't know which one was more efficient/favored.